### PR TITLE
fix: Ensure linear volume subtraction does not go below zero 

### PR DESCRIPTION
**Pull request by foxzool**
fix: [Ensure linear volume subtraction does not go below zero
](https://github.com/bevyengine/bevy/issues/19417)

## Solution
- Clamp the result of linear volume subtraction to a minimum of 0.0
- Add a new test case to verify behavior when subtracting beyond zero

### CODE REVIEW

**Review by janhohenheim — APPROVED** [review]
I think it's worth to discuss whether to keep these impls or not, but while we have them, we should at least not use abs

**Comment by SolarLiner** [comment]
> I think it's worth to discuss whether to keep these impls or not, but while we have them, we should at least not use abs

My recommendation would be for this PR to remove the impls, and mark this a breaking change in case people have accidentally/intentionally relied on this impl existing.

**Comment by foxzool** [comment]
removed the impls   and replace add with increase_by_percentage

File: crates/bevy_audio/src/volume.rs [hunk 1]
```diff
@@ -155,105 +155,195 @@ impl Volume {
...
+    /// let decreased = volume.decrease_by_percentage(50.0);
+    /// assert_eq!(decreased.to_linear(), 0.5);
+    /// ```
+    pub fn decrease_by_percentage(&self, percentage: f32) -> Self {
+        let factor = 1.0 - (percentage / 100.0).clamp(0.0, 1.0);
+        self.adjust_by_linear(factor)
     }
-}
-
-impl core::ops::Mul<Self> for Volume {
```
Comment (SolarLiner) [inline crates/bevy_audio/src/volume.rs]: Mul and Div impls didn't need to be removed, they are strictly equivalent to `adjust_by_linear` and `adjust_by_decibel` depending on the variant of the right-hand side

File: crates/bevy_audio/src/volume.rs [hunk 2]
```diff
@@ -155,57 +155,138 @@ impl Volume {
...
+    ///
+    /// # Examples
+    /// ```
+    /// use bevy_audio::Volume;
+    ///
+    /// let volume = Volume::Linear(0.5);
+    /// let adjusted = volume.adjust_to_linear_target(1.0);
+    /// assert_eq!(adjusted.to_linear(), 1.0);
+    /// ```
+    pub fn adjust_to_linear_target(&self, target_linear: f32) -> Self {
```
Comment (SolarLiner) [inline crates/bevy_audio/src/volume.rs]: This looks like it's equivalent to simply returning `Volume::Linear(target_linear)`

File: crates/bevy_audio/src/volume.rs [hunk 3]
```diff
@@ -155,57 +155,138 @@ impl Volume {
...
+    ///
+    /// # Examples
+    /// ```
+    /// use bevy_audio::Volume;
+    ///
+    /// let volume = Volume::Decibels(-6.0);
+    /// let adjusted = volume.adjust_to_decibel_target(0.0);
+    /// assert!((adjusted.to_decibels() - 0.0).abs() < 0.01);
+    /// ```
+    pub fn adjust_to_decibel_target(&self, target_db: f32) -> Self {
```
Comment (SolarLiner) [inline crates/bevy_audio/src/volume.rs]: This doesn't look too useful either, since you can just `Volume::Decibel(target_db)` directly

**Comment by SolarLiner** [comment]
@alice-i-cecile marked this as breaking because it removes the Add/Sub impls on `Volume`, don't know if that's okay

**Comment by janhohenheim** [comment]
Let's just make sure nothing calls `abs` :eyes:

**Comment by SolarLiner** [comment]
> We should keep the Add and Sub impls intact, and just call the corresponding methods in them + document exactly what they do.

I disagree. The impls as they stand are wrong, and changing them to call "the right methods" would make them the same as the Mul and Div impls, which in turn would make it more confusing than it is worth IMHO. To me, this justifies the breaking change.

Besides, changing their implementation is still a breaking change in that they will have fundamentally different behaviors which which will break existing code.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Alright, I'm convinced. This still needs a migration guide before I can merge it though: see the bot comment above for advice :)

File: release-content/migration-guides/remove_the_add_sub_impls_on_volume.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
+---
+title: remove the Add/Sub impls on Volume
+pull_requests: [ 19423 ]
+---
+
+Linear volumes are like percentages, and it does not make sense to add or subtract percentages.
```
Comment (janhohenheim) [inline release-content/migration-guides/remove_the_add_sub_impls_on_volume.md:6]:
```suggestion
Linear volumes are like percentages, and it does not make sense to add or subtract percentages.
As such, use the new `increase_by_percentage` function instead of addition or subtraction.
```

File: crates/bevy_audio/src/volume.rs [hunk 5]
```diff
@@ -155,57 +155,95 @@ impl Volume {
...
+    /// # Examples
+    /// ```
+    /// use bevy_audio::Volume;
+    ///
+    /// let current = Volume::Linear(1.0);
+    /// let target = Volume::Linear(0.0);
+    /// let faded = current.fade_towards(target, 0.5);
+    /// assert_eq!(faded.to_linear(), 0.5);
+    /// ```
+    pub fn fade_towards(&self, target: Volume, factor: f32) -> Self {
```
Comment (janhohenheim) [inline crates/bevy_audio/src/volume.rs:240]: wonder why this is not implementing `StableInterpolate`

**Review by alice-i-cecile — APPROVED** [review]
